feat: implement custom backend metrics for NIM

keivenchang opened the pull request:
#### Overview:

Implement metrics polling from custom backends (like NIM) that expose their own metrics endpoints. Metrics are collected periodically via background task and exposed through Prometheus gauges.

#### Details:

- @ryan-lempka, note that I added example codes to get you started more easily: launch the Python mock FE+Mock NIM BE together, or the Dynamo FE+Mock NIM BE together. It's in the README.md.
- Add custom backend metrics polling with configurable interval via --custom-backend-metrics-endpoint and --custom-backend-metrics-polling-interval flags
- Implement automatic mode detection: tries static (NATS-only) first, falls back to dynamic (etcd) mode
- Create CustomBackendMetricsRegistry for dynamic Prometheus gauge registration as metrics are discovered
- Add background polling task with proper cancellation and error handling
- Include complete examples: mock_nim_backend.py and mock_nim_frontend.py demonstrating both static and dynamic modes
- All metrics prefixed with dynamo_backend (hard coded) for naming consistency

#### Where should the reviewer start?

- lib/llm/src/http/service/custom_backend_metrics.rs - core polling logic and metrics registry
- lib/llm/src/entrypoint/input/http.rs - DRT mode detection and polling task lifecycle
- examples/custom_backend/nim/README.md - usage examples and configuration

#### Related Issues:

DIS-606

/coderabbit profile chill

keivenchang committed:
feat: add NIM backend metrics polling support

Add polling and on-demand metrics collection from NIM backends:
- Background polling: --nim-metrics-polling-interval-seconds (default: 0=disabled)
- On-demand polling: --nim-metrics-on-demand flag
- Environment variables: NIM_METRICS_POLLING_INTERVAL_SECONDS, NIM_METRICS_ON_DEMAND
- New lib/llm/src/http/service/nim.rs with NimMetricsCollector
- Example mock server in examples/custom_backend/nim/

TODO: Remove once NIM adopts Dynamo backend components natively.

feat: add custom backend metrics polling with static/dynamic mode support

Implement metrics polling from custom backends (like NIM) that expose their
own metrics endpoints. Metrics are collected periodically and exposed through
Prometheus gauges prefixed with dynamo_backend.

Key features:
- Automatic adaptation between static (NATS-only) and dynamic (etcd) modes
- Background polling task with configurable interval
- Dynamic gauge registration as metrics are discovered
- Fallback routing: attempts static mode first, then dynamic mode
- Full example implementation with mock backend and frontend

DEPRECATED: To be removed after custom backends migrate to Dynamo backend
metrics infrastructure.

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

fix: correct type annotation for handle_stats_request return type

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang commented:
Just rebased, no change. Now working on CodeRabbit comments.

nnshah1 commented on components/src/dynamo/frontend/main.py:
```diff
@@ -207,13 +209,32 @@ def parse_args():
+        "--nim-metrics-on-demand",
+        action="store_true",
+        default=os.environ.get(NIM_METRICS_ON_DEMAND_ENV_VAR, "false").lower()
+        in ("true", "1", "yes", "y"),
```
if we have more than one use for this- would recommend a constant "BOOL_TRUTH_VALUES" or similar

nnshah1 commented on lib/llm/src/http/service/nim.rs:
```diff
@@ -0,0 +1,360 @@
+
+    // Call the runtime_stats endpoint with empty request (with timeout)
+    let mut response_stream =
+        match tokio::time::timeout(Duration::from_secs(2), router.random(String::new().into()))
```
can we use direct instead of random here? do we know the explicit instance id - or is that dynamic?

ryan-lempka replied:
Should not matter @keivenchang

ryan-lempka commented on components/src/dynamo/frontend/main.py:
```diff
@@ -207,13 +209,32 @@ def parse_args():
         help="Start KServe gRPC server.",
     )
+    parser.add_argument(
+        "--nim-metrics-polling-interval-seconds",
```
@keivenchang do you think naming should be NIM specific? There may be use cases for others who want to implement their own dynamo workers as well.

keivenchang replied:
I think NIM worker/BE is the first of its kind that is not a Dynamo worker right? I added this as an exception, as this code is temporary.

Dynamo workers adding new metrics, would be using the MetricsRegistry to add metrics.

ryan-lempka replied:
@keivenchang copy that

ryan-lempka reviewed (approved):
One non-blocking comment on naming choice but otherwise LGTM

nnshah1 commented on lib/llm/src/http/service/nim.rs:
```diff
@@ -0,0 +1,360 @@
+
+/// Dynamic NIM metrics registry that maintains gauges for metrics discovered at runtime
+///
+/// Note: These gauges are copies of metrics (gauges, counters, etc.) from the NIM backend.
```
are we descided to avoid counters in this version?

nnshah1 commented on examples/custom_backend/nim/mock_nim_server.py:
```diff
@@ -0,0 +1,84 @@
+        "metrics": {
+            "gauges": {
+                "kv_cache_usage_perc": round(kv_cache_usage, 2),
+                "gpu_utilization_perc": round(gpu_utilization, 1),
```
@ryan-lempka , @keivenchang - are gpu_utilization / active_requests / memory_used_gb - for testing only? are we only using 'kv_cache_usage_perc' - would it make sense to have a known list of metrics instead of converting all?

nnshah1 commented on examples/custom_backend/nim/mock_nim_server.py:
```diff
@@ -0,0 +1,84 @@
+                "kv_cache_usage_perc": round(kv_cache_usage, 2),
+                "gpu_utilization_perc": round(gpu_utilization, 1),
+                "active_requests": active_requests,
+                "memory_used_gb": round(12.5 + random.random() * 2, 2),
```
One other potential, could nim llm backend export worker metrics using the same endpoint as other backends:

WorkerMetricsPublisher

https://github.com/ai-dynamo/dynamo/blob/81162dfeb95adf98d64270ad18a9a4a323e036d8/lib/bindings/python/src/dynamo/_core.pyi#L418

In that way the frontend could poll 'load_metrics' and have a common struct for any dynamo backend - since these are already exported / published. - may be too intrusive to nim llm backend - but a thought.

ryan-lempka reviewed (approved):
A few minor comments but overall LGTM.

keivenchang committed:
feat: improve custom backend metrics validation and consistency

- Add validation to reject negative polling intervals in frontend
- Cap custom backend gauges at 100 to prevent unbounded growth
- Replace hardcoded 'dynamo_backend' prefix with name_prefix::COMPONENT
- Document that all custom backend metrics use 'dynamo_component_' prefix
- Add comprehensive documentation in code comments and README

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
Remove unused nim_on_demand and nim_state parameters from metrics router

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
Merge branch 'main' into keivenchang/DIS-606__poll-NIM-backend-for-metrics

keivenchang committed:
add comment why graceful shutdown is not necessary

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
Merge branch 'main' into keivenchang/DIS-606__poll-NIM-backend-for-metrics

keivenchang committed:
Abort custom backend metrics polling task during shutdown

- Keep JoinHandle alive so task can be explicitly aborted before runtime shutdown
- Add documentation explaining shutdown behavior and why CancellationToken is not used
- Fix README example command for dynamic mode

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>
